src/update_agent/actor.rs

```diff
@@ -73,52 +74,90 @@ impl Message for RefreshTick {
-            UpdateAgentState::UpdateStaged((release, _)) => {
-                let update = release.clone();
-                self.tick_finalize_update(update)
-            }
-            UpdateAgentState::UpdateFinalized(release) => {
-                let update = release.clone();
-                self.tick_end(update)
+    fn handle(&mut self, _msg: RefreshTick, _ctx: &mut Self::Context) -> Self::Result {
+        // Acquire RwLock to access state.
+        let lock = Arc::clone(&self.state);
+        // We need a clone of `info` because we need to move it into futures to ensure a
+        // long enough lifetime.
+        let update_agent_info = self.info.clone();
+        let state_action = async move {
+            let mut agent_state_guard = lock.write_owned().await;
+            // Consider `LAST_REFRESH` time to be when lock is acquired.
+            let tick_timestamp = chrono::Utc::now();
+            LAST_REFRESH.set(tick_timestamp.timestamp());
+
+            trace!("update agent tick, current state: {:?}", *agent_state_guard);
+            let prev_state = (*agent_state_guard).clone();
+
+            let modify_state_outcome = match &prev_state {
+                UpdateAgentState::StartState => {
+                    update_agent_info
+                        .tick_initialize(&mut *agent_state_guard)
+                        .await
+                }
+                UpdateAgentState::Initialized => {
+                    update_agent_info
+                        .tick_report_steady(&mut *agent_state_guard)
+                        .await
+                }
+                UpdateAgentState::ReportedSteady => {
+                    update_agent_info
+                        .tick_check_updates(&mut *agent_state_guard)
+                        .await
+                }
+                UpdateAgentState::NoNewUpdate => {
+                    update_agent_info
+                        .tick_check_updates(&mut *agent_state_guard)
+                        .await
+                }
+                UpdateAgentState::UpdateAvailable((release, _)) => {
+                    let update = release.clone();
+                    update_agent_info
+                        .tick_stage_update(&mut *agent_state_guard, update)
+                        .await
+                }
+                UpdateAgentState::UpdateStaged((release, _)) => {
+                    let update = release.clone();
+                    update_agent_info
+                        .tick_finalize_update(&mut *agent_state_guard, update)
+                        .await
+                }
+                UpdateAgentState::UpdateFinalized(release) => {
+                    let update = release.clone();
+                    update_agent_info
+                        .tick_end(&mut *agent_state_guard, update)
+                        .await
+                }
+                UpdateAgentState::EndState => Ok(()),
+            };
+
+            if modify_state_outcome.is_err() {
```

Comment:
It looks like most of the `tick_*` functions are internally handling error logging, so the error type here is not really used (in fact, it's a `()`). Do you think it would be possible to drop the `Result` at all? That would probably allow factoring our the `.await` too, I think.

Reply:
Ah right, there are actually only a couple `tick`-functions that even make use of this `Result`. I was originally thinking to make minimal changes to the areas that didn't _have_ to be changed. Taking another look to see if not returning `Result` is doable.

Reply:
Follow up to this here https://github.com/coreos/zincati/pull/610